### PR TITLE
Generifying SQL sources and time series

### LINKED ISSUE #473: Generify SQL source classes
As part of #467, extract abstract functionality used for SQL inputs so that they can be reused.

### PR DESCRIPTION
**Pull request by sebastian-peter**
Resolves #473

### CODE REVIEW

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSourceIT.groovy [hunk 1]
```diff
@@ -67,7 +66,7 @@ class CsvTimeSeriesSourceIT extends Specification implements CsvTestDataMeta {
 
 	def "A csv time series source is able to return a time series for a period of interest"() {
 		given:
-		def interval = new ClosedInterval(TimeUtil.withDefaults.toZonedDateTime("2020-01-01 00:15:00"), TimeUtil.withDefaults.toZonedDateTime("2020-01-01 00:15:00"))
+		def interval = new ClosedInterval(TIME_15MIN, TIME_15MIN)
```
Comment (t-ober) [inline src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSourceIT.groovy:69]:
```suggestion
		def interval = new ClosedInterval(TIME_00MIN, TIME_15MIN)
```

**Review by t-ober — CHANGES_REQUESTED** [review]
Nice stuff. Just minor things.

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSourceIT.groovy (hunk 1 above)
Comment (sebastian-peter) [reply src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSourceIT.groovy:69]: That would change the original test. Maybe you can explain why this is better?

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java [hunk 2]
```diff
@@ -26,6 +29,15 @@
...
 public class CsvTimeSeriesSource<V extends Value> extends CsvDataSource
     implements TimeSeriesSource<V> {
+  private static final EnumSet<ColumnScheme> acceptableSchemes =
+      EnumSet.of(
+          ACTIVE_POWER,
+          APPARENT_POWER,
+          ENERGY_PRICE,
+          APPARENT_POWER_AND_HEAT_DEMAND,
+          ACTIVE_POWER_AND_HEAT_DEMAND,
+          HEAT_DEMAND);
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/java/edu/ie3/datamodel/io/source/csv/CsvTimeSeriesSource.java:39]: That's a good point. I had placed this field inside TimeSeriesSource, but Codacy complained about it. Found a better way to do it (using a function) that I'm now including here as well.

File: src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class SqlDataSource<T> {
+
+  protected static final Logger log = LoggerFactory.getLogger(SqlDataSource.class);
+
+  private final SqlConnector connector;
+
+  public SqlDataSource(SqlConnector connector) {
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java:29]: Code Smell: Constructor of abstract classes should be protected as only their sublasses can be instantiated.

File: src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    try (PreparedStatement ps = connector.getConnection().prepareStatement(query)) {
+      addParams.addParams(ps);
+
+      ResultSet resultSet = ps.executeQuery();
+      List<Map<String, String>> fieldMaps = connector.extractFieldMaps(resultSet);
+
+      return fieldMaps.stream()
+          .map(this::createEntity)
+          .flatMap(o -> o.map(Stream::of).orElseGet(Stream::empty))
+          .collect(Collectors.toList());
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/io/source/sql/SqlDataSource.java:107]: Code smell: As of Java 17 you can use Stream.toList()

**Review by t-ober — CHANGES_REQUESTED** [review]
Two code smells we can get rid of otherwise looks good!

**Review by t-ober — APPROVED** [review]

**Pull request merged**